Stack generic AIR constraints structure

0xKanekiKen opened the pull request:
## Describe your changes
This PR introduces a framework of stack operation's generic AIR constraints. New composite flags are introduced for generic constraints of all the stack operations.

## Checklist before requesting a review
- [x] Repo forked and branch created from `next` according to naming convention.
- [x] Commit messages and codestyle follow [conventions](./CONTRIBUTING.md).
- [x] Relevant issues are linked in the PR description.
- [x] Tests added for new functionality.
- [x] Documentation/comments updated according to changes.

0xKanekiKen commented on air/src/stack/mod.rs:
```diff
@@ -92,6 +128,36 @@ pub fn enforce_unique_constraints<E: FieldElement>(
+    // enforces all the stack items post the second elements get copied over onto the next frame.
+    for i in 2..16 {
+        result[constraint_offset] =
+            op_flag.no_change_after_1() * are_equal(frame.stack_item(i), frame.stack_item_next(i));
```
I've handled it using a for loop in the `enforce_generic_function`. This might not be optimal as we might have to update the range or introduce similar loops while introducing new operations. It's better if we handle it in utils.rs in the stack module as you have mentioned in the PR #370. Let me know how do you want to handle it.
Also I was wondering if we should have similar independent generic functions for both left and right shift stack operations?

bobbinth commented:
This approach works, but I think we should do it more holistically. Meaning, we should look at all operations, figure out which flags they require, and then find a good way to merge these flags into a smaller subset. I think doing this incrementally would be very difficult - so, might as well do the whole thing right from the start.

The simplest baseline I can think of is to have 3 flags per stack item - e.g., for item 0, we could have something like `left_shift_0`, `right_shift_0`, `no_change_0`, and similar flags for items 1 through 15. This would result in 48 additional flags. But we should be able to reduce this number considerably.

For example, I don't think there is an instruction which could cause a left shift of item 15 without also doing a left shift for item 14. So, these two can be combined together.

Another aspect of this is that we could probably compute the flags much more efficiently as compared to just summing up flags for relevant operations. This is because many operations which affect the stack in similar ways have common prefixes, and thus, we can use intermediate values (which we compute anyway) to simplify flag computations.

0xKanekiKen commented:
Thanks for the review! After going through all the operations, I've come up with the following composite flags for generic stack constraints which I think covers everything (except `end` control flow operation where the stack state depends on whether the previous block was a loop or not). In 25 additional composite flags we can express all the generic stack constraint in Miden VM.

# Stack composite flags 

## No shift composite flags

$$
`no\_change\_0` = f_{noop} + f_{u32assert2} + f_{mpverify} + f_{join} + f_{span} + f_{respan} + f_{halt}
$$

$$
`no\_change\_1` = `no\_change\_0` + f_{eqz} + f_{neg} + f_{inv} + f_{incr} + f_{not} + f_{neg} + f_{fmpadd} + f_{mload}
$$

$$
`no\_change\_2` = `no\_change\_1` + f_{swap} + f_{u32add} + f_{u32sub} + f_{u32mul} + f_{u32div}
$$

$$
`no\_change\_3` = `no\_change\_2` + f_{movup2} + f_{movdn2}
$$

$$
`no\_change\_4` = `no\_change\_3` + f_{movup3} + f_{movdn3} + f_{readw} + f_{mrupdate} + f_{swapw2} + f_{swapw3}
$$

$$
`no\_change\_5` = `no\_change\_4` + f_{movup4} + f_{movdn4}
$$

$$
`no\_change\_6` = `no\_change\_5` + f_{movup5} + f_{movdn5}
$$

$$
`no\_change\_7` = `no\_change\_6` + f_{movup6} + f_{movdn6}
$$

$$
`no\_change\_8` = `no\_change\_7` - f_{swapw2} + f_{movup7} + f_{movdn7} + f_{swapw}
$$

$$
`no\_change\_9` = `no\_change\_8` + f_{movup8} + f_{movdn8}
$$

$$
`no\_change\_10` = `no\_change\_11` = `no\_change\_9`
$$

$$
`no\_change\_12` = `no\_change\_9` - f_{swapw3}+ f_{swapw2} + f_{rpperm}
$$

$$
`no\_change\_15` = `no\_change\_14` = `no\_change\_13` = `no\_change\_12`
$$

## Left shift composite flags

$$
`left\_shift\_1` = f_{assert} + f_{drop} + f_{mstore} + f_{fmpupdate} + f_{mstorew} + f_{split} + f_{loop} + f_{repeat}
$$

$$
`left\_shift\_2` = `left\_shift\_1` + f_{eq} + f_{add} + f_{mul} + f_{add} + f_{or} + f_{u32and} + f_{u32or}
$$

$$
`left\_shift\_3` = `left\_shift\_2` + f_{cswap} + f_{u32add3} + f_{u32madd}
$$

$$
`left\_shift\_4` = `left\_shift\_3`
$$

$$
`left\_shift\_5` = `left\_shift\_3` + f_{mloadw}
$$

$$
`left\_shift\_8` = `left\_shift\_7` = `left\_shift\_6` = `left\_shift\_5`
$$

$$
`left\_shift\_9` = `left\_shift\_5` + f_{cswapw}
$$

$$
`left\_shift\_15` = `left\_shift\_14` = `left\_shift\_13` = `left\_shift\_12` = `left\_shift\_11` = `left\_shift\_10` = `left\_shift\_9`
$$

## Right shift composite flags

$$
`right\_shift\_0` = \sum (f_{movupn} + f_{movdnn} + f_{dupn}) + f_{pad} + f_{push} + f_{sdepth} + f_{read}
$$

$$
`right\_shift\_1` = `right\_shift\_0` + f_{u32split}
$$

$$
`right\_shift\_2` = `right\_shift\_1` - f_{movup2} - f_{movdn2}
$$

$$
`right\_shift\_3` = `right\_shift\_2` - f_{movup3} - f_{movdn3}
$$

$$
`right\_shift\_4` = `right\_shift\_3` - f_{movup4} - f_{movdn4}
$$

$$
`right\_shift\_5` = `right\_shift\_4` - f_{movup5} - f_{movdn5}
$$

$$
`right\_shift\_6` = `right\_shift\_5` - f_{movup6} - f_{movdn6}
$$

$$
`right\_shift\_7` = `right\_shift\_6` - f_{movup7} - f_{movdn7}
$$

$$
`right\_shift\_8` = `right\_shift\_7` - f_{movup8} - f_{movdn8}
$$

$$
`right\_shift\_15` = `right\_shift\_14` = `right\_shift\_13` = `right\_shift\_12` = `right\_shift\_11` = `right\_shift\_10` = `right\_shift\_9` = `right\_shift\_8`
$$

bobbinth commented:
The approach is right, but a few notes:

1. We probably should cache the flag values. Otherwise, when we call something like `no_change_12()` we also be recomputing all the values which came before it, and we won't realize as much improvement as we could have.
2. Computation of `right_shift_0()` could be simplified quite a bit. Notice that most right shift operations have prefix `011`. We compute value of this prefix when we compute operation flags, and thus should be able to get it for free.
  a. You might be able to apply similar optimizations in other places, but I didn't check.
3. We should also add other composite flags as described [here](https://maticnetwork.github.io/miden/design/stack/op_constraints.html#composite-flags).
4. Regarding `END` operation, we can use the value of one of the helper registers to determine whether we are exiting a `LOOP` block or not (see [here](https://maticnetwork.github.io/miden/design/decoder/main.html#end-operation)).

Lastly, would be good to know total number of field operations to compute these flags.

0xKanekiKen commented:
I've refactored the way the flags were earlier generated in order to reduce the field operations as much as possible.
1) All the composite flags have been cached in the commit [f5e91a0](https://github.com/maticnetwork/miden/pull/392/commits/f5e91a0953938c7a6894a70e7a77a01536b30fa4).
2) I've tried to simplify the computation of composite flags by introducing as helper flags. 
3) Added.
4) End has been incorporated.  

Lastly, the computation now is taking only ~243 field operations. The benckmark is as follows: 

compute_op_flags/op_flags                        
                        time:   [241.12 ns 241.44 ns 241.81 ns]
Found 8 outliers among 100 measurements (8.00%)
  1 (1.00%) low mild
  4 (4.00%) high mild
  3 (3.00%) high severe

Interestingly it's now taking less time to run than [earlier](https://github.com/maticnetwork/miden/pull/380#issuecomment-1234908087) when it has less field operation. I'd need to investigate it.

bobbinth commented on air/src/stack/op_flags/mod.rs:
```diff
@@ -20,6 +20,15 @@ pub const NUM_DEGREE_6_OPS: usize = 16;
 /// Total number of degree 4 operations in the VM.
 pub const NUM_DEGREE_4_OPS: usize = 8;
 
+/// Total number of no shift composite flags in the VM.
+pub const NUM_NO_SHIFT_FLAGS: usize = 16;
+
+/// Total number of left shift composite flags in the VM.
+pub const NUM_LEFT_SHIFT_FLAGS: usize = 16;
+
+/// Total number of right shift composite flags in the VM.
+pub const NUM_RIGHT_SHIFT_FLAGS: usize = 16;
```
Do these (and other) constants in this file need to be public?

Also, do we need to use 16 values for all flags? I thought we could use fewer, at least for some. We could still have all the required methods going up to 16, but the redundant ones could read values from the same places in the array.

0xKanekiKen replied:
> Do these (and other) constants in this file need to be public? 

I'll make them private. They don't need to be public.

> Also, do we need to use 16 values for all flags? I thought we could use fewer, at least for some. We could still have all the required methods going up to 16, but the redundant ones could read values from the same places in the array.

They are still redundant. I've have populated their values with the useful ones. This way we don't need to keep track of the redundant ones with the useful ones and only need to do it in `generate_op_flags` helper function. For eg. `no_change_14` is a redundant flag and uses `no_change_12` flag values. While generating the flags, I've set the 14th index to no_change_12 flag and defined a method `no_change_14` which fetches from the 14th index of no_change_flags.

bobbinth replied:
I guess my thinking was that we won't lose any efficiency but would reduce the number of values in the array. For example, if no change 12, 13, 14, and 15 are the same, we could set `NUM_NO_SHIFT_FLAGS` to 13. And then the relative methods could work as follows:
```Rust
#[inline(always)]
pub fn no_change_12(&self) -> E {
    self.no_shift_flags[12]
}

#[inline(always)]
pub fn no_change_13(&self) -> E {
    self.no_shift_flags[12]
}

#[inline(always)]
pub fn no_change_14(&self) -> E {
    self.no_shift_flags[12]
}

#[inline(always)]
pub fn no_change_15(&self) -> E {
    self.no_shift_flags[12]
}
```

bobbinth commented on air/src/stack/op_flags/mod.rs:
```diff
@@ -62,12 +79,28 @@ impl<E: FieldElement> OpFlags<E> {
     pub fn new(frame: &EvaluationFrame<E>) -> Self {
-        let (degree_6_op_flags, degree_4_op_flags) = set_op_flags_degree_4_and_6(frame);
+        let (
+            degree_7_op_flags,
+            degree_6_op_flags,
+            degree_4_op_flags,
+            no_change_flags,
+            left_shift_at_flags,
+            right_shift_at_flags,
+            left_shift_flag,
+            right_shift_flag,
+            control_flow_flag,
+        ) = generate_op_flags(frame);
 
         Self {
-            degree7_op_flags: set_op_flags_degree_7(frame),
+            degree7_op_flags: degree_7_op_flags,
             degree6_op_flags: degree_6_op_flags,
             degree4_op_flags: degree_4_op_flags,
+            no_shift_flags: no_change_flags,
+            left_shift_flags: left_shift_at_flags,
+            right_shift_flags: right_shift_at_flags,
+            left_shift_flag,
+            right_shift_flag,
+            control_flow_flag,
         }
```
This code seems a bit redundant: should we just move the content of `generate_op_flags` into the constructor here?

0xKanekiKen replied:
But then it will make the constructor messier. Let me know if we should move it in the constructor.

bobbinth replied:
I think that's fine - in the end it is the same function just located in a different place in the file.

bobbinth commented on air/src/stack/op_flags/mod.rs:
```diff
@@ -501,101 +920,147 @@ impl<E: FieldElement> EvaluationFrameExt<E> for &EvaluationFrame<E> {
+
+    // flag of prefix when the first 4 elements in op_bits are `1000`. Caching the result
+    // for the compution of no_change_2. It's a flag for u32 arithmetic operations.
+    let one_zero_zero_zero_flag = degree_7_intermediate_values[0] * not_3 * frame.op_bit(6);
```
nit: it might make it a bit more readable if we use numbers instead of words. For example, instead of `one_zero_zero_zero_flag` we could use `f1000`, instead of `zero_one_zero_zero_flag` we could use `f0100` etc.

bobbinth commented on air/src/stack/op_flags/mod.rs:
```diff
@@ -501,101 +920,147 @@ impl<E: FieldElement> EvaluationFrameExt<E> for &EvaluationFrame<E> {
+    E,
+) {
+    // intermediary array to cache the value of intermediate flags.
+    let mut degree_7_intermediate_values = [E::ZERO; NUM_DEGREE_7_OPS];
     let mut degree_6_intermediate_values = [E::ZERO; NUM_DEGREE_6_OPS];
     let mut degree_4_intermediate_values = [E::ZERO; NUM_DEGREE_4_OPS];
```
nit: maybe we should change the names of these to something like `degree7_op_flags`, `degree6_op_flags` etc. Would make it much less verbose and I don't think it would confuse things too much.

bobbinth reviewed (changes requested):
Looks good! Thank you! I've added a few small comments inline.

Also, we should also add the code which actually uses these flags to enforce generic constraints.

bobbinth commented on air/src/stack/mod.rs:
```diff
@@ -91,6 +116,206 @@ pub fn enforce_unique_constraints<E: FieldElement>(
+    result[0] = op_flag.no_change_0() * are_equal(frame.stack_item(0), frame.stack_item_next(0));
+
+    // enforces the 2nd item of the stack is copied over.
+    result[1] = op_flag.no_change_1() * are_equal(frame.stack_item(1), frame.stack_item_next(1));
```
There is one other way to do this which should reduce the number of constraints 3x. Specifically, we should be able to combine constraints for the same stack item into a single expression:

```
let expected_next_item_1 = op_flag.no_change_1() * frame.stack_item(1)
                         + op_flag.left_shift_2() * frame.stack_item(2)
                         + op_flag.right_shift_0() * frame.stack_item(0);
result[1] = are_equal(frame.stack_item_next(1), expected_next_item_1);
```

0xKanekiKen replied:
Yes, indeed! Thanks, I missed this. I've added it in the commit [bc6237d](https://github.com/maticnetwork/miden/pull/392/commits/bc6237dc9f64b13234fc92e1ef964899d3ed2913).

bobbinth commented on air/src/stack/mod.rs:
```diff
@@ -23,6 +24,20 @@ const B1_COL_IDX: usize = B0_COL_IDX + 1;
+
+/// The degrees of constraints in the generic stack operations.
+pub const CONSTRAINT_DEGREES: [usize; NUM_GENERIC_CONSTRAINTS] = [
+    // Each of the operation are being multiplied with a flag which can be at max of 7 degree.
+    // There are 3 ways in which an operation can effect the stack. It can shift each stack item
+    // to the right or left or doesn't shift it at all. Given only one of the three state transition
+    // is possible, these 3 transition of an item can be combined into one constraint. Therefore, there
+    // will be 16 generic stack constraints encompassing all the 16 stack items. 7 is to each one of
+    // them.
```
nit: I'd probably move this into the doc comment above the constant declaration. Also, last sentence is a bit confusing ("7 is to each...") or at least I didn't understand what it means.

bobbinth commented on air/src/stack/mod.rs:
```diff
@@ -23,6 +24,20 @@ const B1_COL_IDX: usize = B0_COL_IDX + 1;
 pub const NUM_ASSERTIONS: usize = 2 * MIN_STACK_DEPTH + 2;
 
+/// The number of generic constraints in the stack operations.
+pub const NUM_GENERIC_CONSTRAINTS: usize = 16;
```
nit: I'm wonder if the word `generic` is a good one to use here. "generic" has a special meaning in Rust and other languages, maybe a different word would be better - e.g., "general" or maybe "common" or something like that.

bobbinth commented on air/src/stack/mod.rs:
```diff
@@ -91,6 +117,142 @@ pub fn enforce_unique_constraints<E: FieldElement>(
+    result[0] = are_equal(frame.stack_item_next(0) * flag_sum_0, expected_next_item_0);
+
+    // enforces constraint on the 2nd element in the stack in the next trace.
+    let flag_sum_1 = op_flag.no_change_1() + op_flag.left_shift_2() + op_flag.right_shift_0();
+    let expected_next_item_1 = op_flag.no_change_1() * frame.stack_item(1)
+        + op_flag.left_shift_2() * frame.stack_item(2)
+        + op_flag.right_shift_0() * frame.stack_item(0);
+    result[1] = are_equal(frame.stack_item_next(1) * flag_sum_1, expected_next_item_1);
```
This works - but it results in quite a bit of code - we could probably try to simplify this using a loop. For this, we'd need to change the signature of `left_shift()`, `right_shift()`, and `no_change()` methods to accept an index (similar to how we do it with stack items). Then, the body of the loop would look something like:

```Rust
let flag_sum = op_flag.no_change(i) + op_flag.left_shift(i + 1) + op_flag.right_shift(i - 1);
let expected_next_item = op_flag.no_change(i) * frame.stack_item(i)
        + op_flag.left_shift(i + 1) * frame.stack_item(i + 1)
        + op_flag.right_shift(i - 1) * frame.stack_item(i - 1);
result[i] = are_equal(frame.stack_item_next(i) * flag_sum, expected_next_item);
```

We'd still have to do the first and the last iterations manually, but at least all other iterations would be handled by a loop.

bobbinth replied:
Yep, in this case having all arrays go up to 16 is a better option.

bobbinth commented on air/src/stack/op_flags/tests.rs:
```diff
@@ -121,3 +121,410 @@ fn degree_4_op_flags() {
+        // All the operation flags are generated for the given frame.
+        let op_flags = OpFlags::new(&frame);
+
+        for i in op_flags.no_shift_flags.iter() {
+            assert_eq!(*i, ONE);
+        }
```
nit: instead of dereferencing `i` as `*i`, you could re-write this as:
```Rust
for &i in op_flags.no_shift_flags.iter() {
    assert_eq!(i, ONE);
}
```

bobbinth commented on air/src/stack/op_flags/tests.rs:
```diff
@@ -121,3 +121,410 @@ fn degree_4_op_flags() {
+            assert_eq!(op_flags.control_flow_flag, ZERO);
+        } else if op == Operation::Span || op == Operation::Halt {
+            assert_eq!(op_flags.control_flow_flag, ONE);
+        } else {
+        }
```
nit: instead of leaving the `else` clause empty, I'd do something like:
```Rust
} else {
    unreachable!("unexpected op");
}
```

bobbinth commented on air/src/stack/op_flags/tests.rs:
```diff
@@ -121,3 +121,410 @@ fn degree_4_op_flags() {
+    // All the operation flags are generated for the given frame.
+    let op_flags = OpFlags::new(&frame);
+
+    for (i, ele) in op_flags.no_shift_flags.iter().enumerate() {
+        if i == 0 {
+            assert_eq!(*ele, ZERO);
+        } else {
+            assert_eq!(*ele, ONE);
+        }
+    }
```
nit: a slightly more readable way would be to check the first flag outside of the loop, and the all other ones inside a loop.

bobbinth reviewed (changes requested):
Looks good! I left a few more small comments inline.

bobbinth commented on air/src/stack/op_flags/mod.rs:
```diff
@@ -76,405 +356,530 @@ impl<E: FieldElement> OpFlags<E> {
+
+    /// Returns the flag when the stack operation shifts the flag to the right.
+    #[inline(always)]
+    pub fn right_shift_flag(&self) -> E {
```
nit: let's rename this to just `right_shift` and same for other similar methods.

bobbinth commented on air/src/stack/op_flags/mod.rs:
```diff
@@ -487,6 +892,10 @@ trait EvaluationFrameExt<E: FieldElement> {
+
+    /// Returns the value if the `h5` helper register in the decoder which is set to ONE if the
+    /// exiting block is a `LOOP` block.
+    fn prev_loop(&self) -> E;
```
nit: maybe call this `is_loop` or `was_loop`?

bobbinth commented on air/src/stack/op_flags/mod.rs:
```diff
@@ -61,13 +79,275 @@ impl<E: FieldElement> OpFlags<E> {
+            + degree6_op_flags[13]
+            + degree4_op_flags[6]
+            + degree4_op_flags[7]
+            + end_flag;
```
Is this correct? I see `end_flag` being used to indicate no shift here, and later on to indicate left shift. It seems like it can't be used in both places? Maybe this should actually be `degree4_op_flags[4] * not(frame.prev_loop())`? If this is the case, let's create test to catch this issue.

0xKanekiKen replied:
Thanks! I missed it. I've added two test cases for the end operation covering both the cases.

bobbinth commented on air/src/stack/op_flags/mod.rs:
```diff
@@ -61,13 +79,275 @@ impl<E: FieldElement> OpFlags<E> {
+
+        // flag of END operation. It's effect on stack depends if the current block being executed is
+        // a loop block or not.
+        let end_flag = degree4_op_flags[4] * frame.prev_loop();
```
I think the naming `end_flag` is confusing. I think think what this flag actually means is `shift_left_on_end`, right?

bobbinth reviewed (changes requested):
Thank you! Looks good! I left a few more comments inline.

0xKanekiKen commented on air/src/stack/mod.rs:
```diff
@@ -91,6 +119,40 @@ pub fn enforce_unique_constraints<E: FieldElement>(
+        + op_flag.right_shift(14) * frame.stack_item(14);
+    result[15] = are_equal(frame.stack_item_next(15) * flag_sum, expected_next_item);
+
+    16
```
Done

0xKanekiKen commented on air/src/stack/mod.rs:
```diff
@@ -91,6 +119,40 @@ pub fn enforce_unique_constraints<E: FieldElement>(
+        .iter_mut()
+        .enumerate()
+        .take(NUM_GENERAL_CONSTRAINTS - 1)
+        .skip(1)
```
yes - I've had to add an attribute to suppress it.
